pkg/aws/eks/eks.go

```diff
@@ -0,0 +1,360 @@
+			}
+			spotPrices = append(spotPrices, spotPriceList...)
+			c.NextScrape = time.Now().Add(c.ScrapeInterval)
+		}
+		c.pricingMap = NewStructuredPricingMap()
+		if err := c.pricingMap.GeneratePricingMap(prices, spotPrices); err != nil {
+			return fmt.Errorf("%w: %w", GeneratePricingMapErr, err)
+		}
+	}
+
+	wg := sync.WaitGroup{}
+	wg.Add(len(c.Regions))
+	instanceCh := make(chan []ec2Types.Reservation, len(c.Regions))
+	for _, region := range c.Regions {
+		go func(region ec2Types.Region, profile string) {
+			defer wg.Done()
+			client := c.ec2RegionClient[*region.RegionName]
+			reservations, err := ListComputeInstances(context.Background(), client)
+			if err != nil {
+				log.Printf("error listing instances: %s", err)
+				return
+			}
+			log.Printf("found %d instances in profile:region %s:%s", len(reservations), profile, *region.RegionName)
+			instanceCh <- reservations
+		}(region, c.Profile)
+	}
+	go func() {
+		wg.Wait()
+		close(instanceCh)
+	}()
+	c.emitMetricsFromChannel(instanceCh, ch)
+	return nil
+}
+
+func (c *Collector) emitMetricsFromChannel(reservationsCh chan []ec2Types.Reservation, ch chan<- prometheus.Metric) {
+	for reservations := range reservationsCh {
+		for _, reservation := range reservations {
+			for _, instance := range reservation.Instances {
+				clusterName := clusterNameFromInstance(instance)
+				if clusterName == "" {
+					log.Printf("no cluster name found for instance %v", instance)
+					continue
+				}
+				if instance.PrivateDnsName == nil || *instance.PrivateDnsName == "" {
+					log.Printf("no private dns name found for instance %v", instance)
+					continue
+				}
+				if instance.Placement == nil || instance.Placement.AvailabilityZone == nil {
+					log.Printf("no availability zone found for instance %v", instance)
+					continue
+				}
+
+				region := *instance.Placement.AvailabilityZone
+
+				pricetier := "spot"
+				if instance.InstanceLifecycle != ec2Types.InstanceLifecycleTypeSpot {
+					pricetier = "ondemand"
+					// Ondemand instances are keyed based upon their region, so we need to remove the availability zone
+					region = region[:len(region)-1]
+				}
+				price, err := c.pricingMap.GetPriceForInstanceType(region, string(instance.InstanceType))
+				if err != nil {
+					log.Printf("error getting price for instance type %s: %s", instance.InstanceType, err)
+					continue
+				}
+				labelValues := []string{
+					*instance.PrivateDnsName,
+					region,
+					c.pricingMap.InstanceDetails[string(instance.InstanceType)].InstanceFamily,
+					string(instance.InstanceType),
+					clusterName,
+					pricetier,
+					"eks",
+				}
+				ch <- prometheus.MustNewConstMetric(InstanceCPUHourlyCostDesc, prometheus.GaugeValue, price.Cpu, labelValues...)
+				ch <- prometheus.MustNewConstMetric(InstanceMemoryHourlyCostDesc, prometheus.GaugeValue, price.Ram, labelValues...)
+			}
+		}
+	}
+}
+
+func CollectMetrics(_ chan<- prometheus.Metric) error {
+	//TODO implement me
+	panic("implement me")
```

Comment:
?

Reply:
🤔 thought I provided a response. `CollectMetrics` is a leaky abstraction that I'm looking to remove in a future PR. For now I'll return nil since panic'ing seems excessive.